### PR TITLE
Implement safer variant to ensure CDATA end markers result in valid XML

### PR DESCRIPTION
**Pull request by chschu**
## Overview

This PR changes the implementation of `XmlReportWriter. writeCDataSafely(...)` to be even safer by always passing valid CDATA content to `javax.xml.stream.XMLStreamWriter.writeCData(...)`.

The current implementation passes a String to `javax.xml.stream.XMLStreamWriter.writeCData(...)` that might not be valid CDATA content, i.e. it might contain `]]>`. This causes issues if the implementation of that method validates its parameter value.

The PR does not change the produced XML, which is covered by `XmlReportWriterTests.writesValidXmlEvenIfExceptionMessageContainsCData()`.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [X] There are no TODOs left in the code
- [X] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [X] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [X] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [X] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [ ] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

**Comment by marcphilipp** [comment]
@chschu Thanks! Could you please add a test case that demonstrates the issue and ensures we don't introduce it again?

**Comment by chschu** [comment]
@marcphilipp I didn't want to expose more of the `XmlReportWriter` implementation just for the test. The arguments of `XMLStreamWriter.writeCData(...)` cannot be captured directly in a test, because the `XMLOutputFactory` is obtained using a static call.

The test I added checks that the data passed to the underlying `Writer` never contains the problematic substring `]]><![CDATA[`. That shouldn't ever happen if CDATA blocks are written separately.

The CI failure on Windows looks like a file locking issue that has nothing to do with my changes.

File: junit-platform-reporting/src/main/java/org/junit/platform/reporting/legacy/xml/XmlReportWriter.java [hunk 1]
```diff
@@ -51,8 +52,7 @@
  */
 class XmlReportWriter {
 
-	private static final String CDATA_START = "<![CDATA[";
-	private static final String CDATA_END = "]]>";
+	private static final Pattern CDATA_SPLIT_PATTERN = Pattern.compile("(?<=]])(?=>)");
```
Comment (marcphilipp) [inline junit-platform-reporting/src/main/java/org/junit/platform/reporting/legacy/xml/XmlReportWriter.java]: ❓ Why is this pattern so complicated? Is the positive lookbehind and the positive lookahead necessary?

File: platform-tests/src/test/java/org/junit/platform/reporting/legacy/xml/XmlReportWriterTests.java [hunk 2]
```diff
@@ -219,10 +220,44 @@ void writesValidXmlEvenIfExceptionMessageContainsCData() throws Exception {
...
+			}
+
+			@Override
+			public void flush() {
+			}
+
+			@Override
+			public void close() {
+			}
+		};
```
Comment (sbrannen) [inline platform-tests/src/test/java/org/junit/platform/reporting/legacy/xml/XmlReportWriterTests.java]: Please replace that block with the following.

```java
		Writer assertingWriter = new StringWriter() {

			@Override
			public void write(char[] cbuf, int off, int len) {
				assertThat(new String(cbuf, off, len)).doesNotContain("]]><![CDATA[");
			}
		};
```

File: junit-platform-reporting/src/main/java/org/junit/platform/reporting/legacy/xml/XmlReportWriter.java (hunk 1 above)
Comment (sbrannen) [reply junit-platform-reporting/src/main/java/org/junit/platform/reporting/legacy/xml/XmlReportWriter.java]: And if the the positive lookbehind and  positive lookahead are necessary, please add comments in the code to explain exactly what the pattern matches and why.

Comment (marcphilipp) [reply junit-platform-reporting/src/main/java/org/junit/platform/reporting/legacy/xml/XmlReportWriter.java]: @chschu Could you please take a look? We'd like to include this in 5.6 M1.

Comment (chschu) [reply junit-platform-reporting/src/main/java/org/junit/platform/reporting/legacy/xml/XmlReportWriter.java]: Sorry for the delay.

The zero-width assertions are used to simplify the loop in `writeCDataSafely(...)`. Any characters captured by the pattern itself would have to be added to the results of `String.split(...)`.

I added a comment that should clarify this.

Comment (marcphilipp) [reply junit-platform-reporting/src/main/java/org/junit/platform/reporting/legacy/xml/XmlReportWriter.java]: Thanks! 👍

**Comment by marcphilipp** [comment]
Thanks, @chschu! 👍
